horovod/run/elastic/driver.py

```diff
@@ -0,0 +1,283 @@
+
+        if len(self._host_assignments) > 0:
+            # Ensure that at least one previously active host is still assigned, otherwise there is no
+            # way to sync the state to the new workers
+            prev_hosts = self._host_assignments.keys()
+            next_hosts = host_assignments.keys()
+            if not prev_hosts & next_hosts:
+                raise RuntimeError('No hosts from previous set remaining, unable to broadcast state.')
+
+        self._host_assignments = host_assignments
+        self._ordered_available_hosts = ordered_available_hosts
+        self._world_size = len(host_assignments_list)
+        self._rendezvous.httpd.init(host_assignments_list)
+
+        # Rank assignments map from world rank to slot info
+        rank_assignments = {}
+        for slot_info in host_assignments_list:
+            rank_assignments[slot_info.rank] = slot_info
+        self._rank_assignments = rank_assignments
+
+        # Get the newly assigned slots that need to be started
+        pending_slots = [slot_info
+                         for host, slots in self._host_assignments.items()
+                         for slot_info in slots
+                         if (host, slot_info.local_rank) not in active_slots]
+        return pending_slots
+
+    def _get_host_assignments(self):
+        # We need to ensure this list preserves relative order to ensure the oldest hosts are assigned lower ranks.
+        ordered_available_hosts = self._discovered_hosts.filter_available_hosts(self._ordered_available_hosts)
+        known_hosts = set(self._ordered_available_hosts)
+        for host in self.get_available_hosts():
+            if host not in known_hosts:
+                ordered_available_hosts.append(host)
+
+        # Adjust the host assignments to account for added / removed hosts
+        host_list = [hosts.HostInfo(host, self._discovered_hosts.get_slots(host)) for host in ordered_available_hosts]
+        host_assignments_list = hosts.get_host_assignments(host_list, self._min_np, self._max_np)
+        host_assignments = defaultdict(list)
+        for slot_info in host_assignments_list:
+            host_assignments[slot_info.hostname].append(slot_info)
+        return host_assignments, host_assignments_list, ordered_available_hosts
+
+    def _start_worker_processes(self, pending_slots):
+        for slot_info in pending_slots:
+            logging.info('start worker process: {}[{}]'.format(slot_info.hostname, slot_info.local_rank))
+            self._start_worker_process(slot_info)
+
+    def _start_worker_process(self, slot_info):
+        create_worker_fn = self._create_worker_fn
+        shutdown_event = self._shutdown
+        host_event = self._discovered_hosts.get_host_event(slot_info.hostname)
+
+        def run_worker():
+            res = create_worker_fn(slot_info, [shutdown_event, host_event])
+            exit_code, timestamp = res
+            self._handle_worker_exit(slot_info, exit_code, timestamp)
```

Comment:
Image the following situation:
- We have a host with two slots and start the workers.
- Then the first slot becomes unavailable and the first worker gets pre-empted.
- Later a second slot becomes available again. In host assignment, the second local rank will be added to pending slots so that the second local rank will be started again.
- Then we have two worker running with the same slot_info. Both will register their exit codes with the second local rank and none will register a successful exit code for the first local rank.

Reply:
In the current implementation, this won't be a problem because of how aggressive we are at blacklisting hosts for any failure.  In the future, once we flesh out the blacklist logic, I agree, this is something we need to be mindful of.  I will add a task to track the improved blacklisting.

Reply:
#1926